### PR TITLE
Improve TICL Association Map and update TICL event content

### PR DESCRIPTION
**Pull request by felicepantaleo**
This pull request introduces a comprehensive enhancement of the `TICLAssociationMap` class and updates the modules that depend on it to improve flexibility, readability, and performance for association handling within the TICL framework.

- Refactoring of `TICLAssociationMap`:
  - Introduced new wrapper types `FractionType` and `SharedEnergyType` to differentiate between fraction and shared energy values more clearly.
  - Created a generic `AssociationElement` class to manage indices and associated values, including methods for fractional and shared energy access and accumulation operations.
  - Restructured map definitions to make use of `AssociationElement` for better encapsulation and flexibility in handling different types of associations.
  - Enhanced sort() Method in `AssociationMap`:
    - Tie-Breaker for Reproducibility: The sort() method now includes a tie-breaker based on indices when sorting elements with equal values (e.g., score, fraction, or sharedEnergy). This ensures that the sorting is reproducible and consistent across runs.
    - Custom Comparator Support: An overloaded sort() method has been added to accept a custom comparison function. This provides users with the flexibility to define their own sorting criteria based on specific needs.
  - Initialization of `AssociationElement` with Invalid State:
  - Invalid Default State: Modified the `AssociationElement` class to initialize elements in an invalid state by default. This is achieved by setting the `index_` to `std::numeric_limits<unsigned int>::max()` and `value_` to `-1.f `for relevant types.
  - Updated various producers and consumers of the `TICLAssociationMap` to work with the enhanced API.
  - Replaced direct vector manipulation with the new `AssociationElement` for consistency and code clarity.
  
Bug fixes:
  - The maps are now sorted by score in ascending order, and the sorting is made reproducible by including the tie-breaker based on indices. This fixes the wrong behavior shown in DQM plots of best associated trackster and second best associated trackster
  - Fixed Access Issues: In HGVHistoProducerAlgo.cc, replaced direct indexing with the .at() method when accessing the scToCpMap. 
  - Fixed insert() in one-to-one maps 
- Validation and Testing:
  - Adjusted validation and monitoring code in modules such as `HGCalValidator` and `TICLCandidateValidator` to align with the new data structures and API changes.
- Make the association maps persistent.
- Include SimTauProducer in the event content for ticlv5

### CODE REVIEW

**Comment by felicepantaleo** [comment]
test parameters:
- workflow_opts= -w upgrade
- workflow = 29896.203, 29691.203, 29691.204
- enable = hlt_p2_integration, hlt_p2_timing

**Comment by felicepantaleo** [comment]
type hgcal

**Comment by rovere** [comment]
Did you expect regression on the score? As far as I can see, this is mostly (only?) evident in the sim2reco since the regression affects purity and efficiency but not fake and merge.

File: Validation/HGCalValidation/scripts/makeHGCalValidationPlots.py [hunk 1]
```diff
@@ -29,9 +29,10 @@
 tracksters = []
 
 def main(opts):
-
+    
     drawArgs={}
     extendedFlag = False
+    ticlVersion = 4
```
[the remark this answers is not in the document]
Comment (felicepantaleo) [reply Validation/HGCalValidation/scripts/makeHGCalValidationPlots.py:35]: the default is still v4

**Comment by felicepantaleo** [comment]
@cmsbuild please test

File: Validation/HGCalValidation/scripts/makeHGCalValidationPlots.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (felicepantaleo) [reply Validation/HGCalValidation/scripts/makeHGCalValidationPlots.py:35]: simTauProducer is already in the event content for ticlv4, it was added to the event content for when the process modifier for ticlv5 is used

**Comment by felicepantaleo** [comment]
> Did you expect regression on the score? As far as I can see, this is mostly (only?) evident in the sim2reco since the regression affects purity and efficiency but not fake and merge.

I expect changes in the entries in non-particlegun events, as some simtracksters were wrongly assigned to CP 0 which is typically among the selected CPs.

**Comment by jfernan2** [comment]
please test
Let's try again

**Comment by jfernan2** [comment]
please test

**Comment by felicepantaleo** [comment]
@cmsbuild please test

**Comment by felicepantaleo** [comment]
@cmsbuild please test

**Comment by felicepantaleo** [comment]
also this test will fail in PU with premix, debugging.

**Comment by felicepantaleo** [comment]
@cmsbuild please test

**Comment by jfernan2** [comment]
+1

**Comment by Moanwar** [comment]
+Upgrade

**Comment by antoniovagnerini** [comment]
+dqm

**Comment by mandrenguyen** [comment]
+1

**Pull request merged**